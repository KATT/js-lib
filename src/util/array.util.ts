import { NotVoid, RecursiveArray, StringIteratee, ValueIteratee } from './lodash.types'

/**
 * Creates an array of elements split into groups the length of size. If collection canâ€™t be split evenly, the
 * final chunk will be the remaining elements.
 *
 * @param array The array to process.
 * @param size The length of each chunk.
 * @return Returns the new array containing chunks.
 *
 * https://lodash.com/docs#chunk
 *
 * Based on: https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_chunk
 */
export function _chunk<T> (array: T[], size = 1): T[][] {
  return array.reduce(
    (arr, item, idx) => {
      return idx % size === 0
        ? [...arr, [item]]
        : [...arr.slice(0, -1), [...arr.slice(-1)[0], item]]
    },
    [] as T[][],
  )
}

/**
 * Polyfill to Array.flat() with depth=1.
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat
 */
export function _flatten<T> (arrays: T[][]): T[] {
  // to flat single level array
  return ([] as T[]).concat(...arrays)
}

/**
 * Recursively flattens a nested array.
 *
 * @param arr The array to recursively flatten.
 * @return Returns the new flattened array.
 *
 * Based on: https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_flattendeep
 */
export function _flattenDeep<T> (arr: RecursiveArray<T>): T[] {
  return Array.isArray(arr)
    ? arr.reduce((a: RecursiveArray<T>, b) => a.concat(_flattenDeep(b as RecursiveArray<T>)), [])
    : ([arr] as any)
}

/**
 * Removes duplicates from given array.
 */
export function _uniq<T> (a: T[]): T[] {
  return [...new Set(a)]
}

/**
 * This method is like `_.uniq` except that it accepts `iteratee` which is
 * invoked for each element in `array` to generate the criterion by which
 * uniqueness is computed. The iteratee is invoked with one argument: (value).
 *
 * @category Array
 * @param arr The array to inspect.
 * @param predicate The iteratee invoked per element.
 * @returns Returns the new duplicate free array.
 * @example
 *
 * _.uniqBy([2.1, 1.2, 2.3], Math.floor);
 * // => [2.1, 1.2]
 *
 * // using the `_.property` iteratee shorthand
 * _.uniqBy([{ 'x': 1 }, { 'x': 2 }, { 'x': 1 }], 'x');
 * // => [{ 'x': 1 }, { 'x': 2 }]
 *
 * Based on: https://stackoverflow.com/a/40808569/4919972
 */
export function _uniqBy<T> (arr: T[], predicate: ValueIteratee<T>): T[] {
  const cb = typeof predicate === 'function' ? predicate : (o: T) => o[predicate as any]

  return [
    ...arr
      .reduce((map, item) => {
        const key = item === null || item === undefined ? item : cb(item)

        if (!map.has(key)) map.set(key, item)

        return map
      }, new Map())
      .values(),
  ]
}

/**
 * const a = [
 *  {id: 'id1', a: 'a1'},
 *  {id: 'id2', b: 'b1'},
 * ]
 *
 * by(a, 'k')
 * // => {
 *   id1: {id: 'id1', a: 'a1'},
 *   id2: {id: 'id2', b: 'b1'},
 * }
 *
 * by(a, k => k.id.toUpperCase())
 * // => {
 *   ID1: {id: 'id1', a: 'a1'},
 *   ID2: {id: 'id2', b: 'b1'},
 * }
 */
export function by<T> (items: T[] = [], predicate: StringIteratee<T>): Record<string, T> {
  const cb: (value: T) => string | undefined =
    typeof predicate === 'function' ? predicate : (item: T) => item[predicate]

  return items.reduce(
    (map, item) => {
      const res = cb(item)
      if (res) map[res] = item
      return map
    },
    {} as Record<string, T>,
  )
}

/**
 * _sortBy([{age: 20}, {age: 10}], 'age')
 * // => [{age: 10}, {age: 20}]
 *
 * Same:
 * _sortBy([{age: 20}, {age: 10}], o => o.age)
 */
export function _sortBy<T> (items: T[], predicate: ValueIteratee<T>): T[] {
  const cb: (value: T) => NotVoid =
    typeof predicate === 'function' ? predicate : (item: T) => item[predicate as string]

  return [...items].sort((_a, _b) => {
    const [a, b] = [_a, _b].map(cb)
    if (typeof a === 'number' && typeof b === 'number') return Math.sign(a - b)
    return String(a).localeCompare(String(b))
  })
}

/**
 * Returns an array with ranges from `from` up to (but not including) `to`
 *
 * @example
 * range(3, 6) // [ 3, 4, 5 ]
 */
export function _range (fromIncl: number, toExcl: number): number[] {
  return Array.from({ length: toExcl - fromIncl }, (_v, k) => k + fromIncl)
}
