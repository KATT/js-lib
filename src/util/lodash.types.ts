export type NotVoid = unknown
export type PropertyName = string | number | symbol
export type PartialShallow<T> = {
  [P in keyof T]?: T[P] extends object ? object : T[P]
}
export type IterateeShorthand<T> = PropertyName | [PropertyName, any] | PartialShallow<T>
export type ValueIteratee<T> = ((value: T) => NotVoid) | IterateeShorthand<T>

export interface RecursiveArray<T> extends Array<T | RecursiveArray<T>> {}

export type Many<T> = T | ReadonlyArray<T>
export type PropertyPath = Many<PropertyName>
