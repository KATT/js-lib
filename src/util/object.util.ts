import { StringMap } from '../types'

export class ObjectUtil {
  // Picks just allowed fields and no other fields
  // pick<T> (obj: any, fields: string[], initialObject = {}): T {
  pick<T> (obj: T, fields?: string[], initialObject: Partial<T> = {}): T {
    if (!obj || !fields || !fields.length) return obj

    const o: any = initialObject

    fields.forEach(k => {
      if (k in obj) o[k] = (obj as any)[k]
    })

    return o
  }

  /**
   * Does NOT mutate (unless "mutate" flag is set)
   * Removes "falsy" value from the object.
   */
  filterFalsyValues<T> (_obj: T, mutate = false): T {
    return this.filterValues(_obj, (k, v) => !!v, mutate)
  }

  filterEmptyStringValues<T> (_obj: T, mutate = false): T {
    return this.filterValues(_obj, (k, v) => v !== '', mutate)
  }

  filterUndefinedValues<T> (_obj: T, mutate = false): T {
    return this.filterValues(_obj, (k, v) => v !== undefined && v !== null, mutate)
  }

  filterValues<T> (_obj: T, predicate: (key: any, value: any) => boolean, mutate = false): T {
    if (!this.isObject(_obj)) return _obj

    const o: any = mutate ? _obj : { ...(_obj as any) }

    Object.keys(o).forEach(k => {
      const keep = predicate(k, o[k])
      if (!keep) delete o[k]
    })

    return o
  }

  transformValues<T> (_obj: T, transformFn: (key: any, value: any) => any, mutate = false): T {
    if (!this.isObject(_obj)) return _obj

    const o: any = mutate ? _obj : { ...(_obj as any) }

    Object.keys(o).forEach(k => {
      o[k] = transformFn(k, o[k])
    })

    return o
  }

  objectNullValuesToUndefined<T> (_obj: T, mutate = false): T {
    return this.transformValues(
      _obj,
      (k, v) => {
        if (v === null) return undefined
        return v
      },
      mutate,
    )
  }

  deepEquals (a: object, b: object): boolean {
    return JSON.stringify(this.sortObjectDeep(a)) === JSON.stringify(this.sortObjectDeep(b))
  }

  /**
   * Deep copy object (by json parse/stringify).
   */
  deepCopy<T> (o: T): T {
    return JSON.parse(JSON.stringify(o))
  }

  isObject (item: any): boolean {
    return (item && typeof item === 'object' && !Array.isArray(item) && item !== null) || false
  }

  isEmptyObject (obj: any): boolean {
    return obj && obj.constructor === Object && Object.keys(obj).length === 0
  }

  // from: https://gist.github.com/Salakar/1d7137de9cb8b704e48a
  mergeDeep (target: any, source: any): any {
    if (this.isObject(target) && this.isObject(source)) {
      Object.keys(source).forEach(key => {
        if (this.isObject(source[key])) {
          if (!target[key]) Object.assign(target, { [key]: {} })
          this.mergeDeep(target[key], source[key])
        } else {
          Object.assign(target, { [key]: source[key] })
        }
      })
    }
    return target
  }

  /**
   * Mutates
   */
  deepTrim (o: any): any {
    if (!o) return o

    if (typeof o === 'string') {
      return o.trim()
    } else if (typeof o === 'object') {
      Object.keys(o).forEach(k => {
        o[k] = this.deepTrim(o[k])
      })
    }

    return o
  }

  private defaultSortFn (a: any, b: any): number {
    return a.localeCompare(b)
  }

  // based on: https://github.com/IndigoUnited/js-deep-sort-object
  sortObjectDeep<T> (o: T): T {
    // array
    if (Array.isArray(o)) {
      return o.map(i => this.sortObjectDeep(i)) as any
    }

    if (this.isObject(o)) {
      const out: any = {}

      Object.keys(o)
        .sort((a, b) => this.defaultSortFn(a, b))
        .forEach(k => {
          out[k] = this.sortObjectDeep((o as any)[k])
        })

      return out
    }

    return o
  }

  // from: https://github.com/jonschlinkert/unset-value
  // mutates obj
  unsetValue (obj: any, prop: string): void {
    if (!this.isObject(obj)) {
      return
    }
    if (obj.hasOwnProperty(prop)) {
      delete obj[prop]
      return
    }

    const segs = prop.split('.')
    let last = segs.pop()
    while (segs.length && segs[segs.length - 1].slice(-1) === '\\') {
      last = segs.pop()!.slice(0, -1) + '.' + last
    }
    while (segs.length && this.isObject(obj)) {
      const k = (prop = segs.shift()!)
      obj = obj[k]
    }
    if (!this.isObject(obj)) return
    delete obj[last!]
  }

  /**
   * Returns object with filtered keys from "exclude" array.
   * E.g:
   * mask({...}, [
   *  'account.id',
   *  'account.updated',
   * ])
   *
   * Pass deepCopy if you want to protect the whole object (not just first level) from mutation.
   */
  mask<T extends object> (_o: T, exclude: string[], deepCopy = false): T {
    let o = { ...(_o as object) }
    if (deepCopy) o = this.deepCopy(o)

    exclude.forEach(e => {
      // eval(`delete o.${e}`)
      this.unsetValue(o, e)
    })
    return o as T
  }

  /**
   * Turns ['a', b'] into {a: true, b: true}
   */
  arrayToHash (a: any[] = []): { [k: string]: boolean } {
    return a.reduce((o, item) => {
      o[item] = true
      return o
    }, {})
  }

  classToPlain<T = any> (obj: any = {}): T {
    return JSON.parse(JSON.stringify(obj))
  }

  getKeyByValue<T = any> (object: any, value: any): T | undefined {
    if (!this.isObject(object)) return
    return Object.keys(object).find(k => object[k] === value) as any
  }

  invertObject<T> (o: any): T {
    const inv: any = {}
    Object.keys(o).forEach(k => {
      inv[o[k]] = k
    })
    return inv
  }

  invertMap<K, V> (m: Map<K, V>): Map<V, K> {
    const inv = new Map<V, K>()
    m.forEach((v, k) => inv.set(v, k))
    return inv
  }

  by<T = any> (items: T[] = [], by: string): StringMap<T> {
    const r: StringMap<T> = {}
    items.forEach((item: any) => {
      if (item[by]) r[item[by]] = item
    })
    return r
  }
}

export const objectUtil = new ObjectUtil()
