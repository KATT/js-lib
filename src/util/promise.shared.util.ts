export interface Deferred<T> {
  promise: Promise<T>
  resolve (a?: T): void
  reject (e?: any): void
}

class PromiseSharedUtil {
  defer<T = void> (): Deferred<T> {
    const deferred = {} as Deferred<T>

    deferred.promise = new Promise((resolve, reject) => {
      deferred.resolve = resolve
      deferred.reject = reject
    })

    return deferred
  }

  async delay (ms: number): Promise<void> {
    return new Promise<void>(r => setTimeout(r, ms))
  }

  async hangingPromise (): Promise<void> {
    return new Promise<void>(r => {})
  }
}

export const promiseSharedUtil = new PromiseSharedUtil()
