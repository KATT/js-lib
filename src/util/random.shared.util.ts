class RandomSharedUtil {
  randomInt (minIncl: number, maxIncl: number): number {
    return Math.floor(Math.random() * (maxIncl - minIncl + 1) + minIncl)
  }
}

export const randomSharedUtil = new RandomSharedUtil()
