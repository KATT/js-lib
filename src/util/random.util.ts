export function randomInt (minIncl: number, maxIncl: number): number {
  return Math.floor(Math.random() * (maxIncl - minIncl + 1) + minIncl)
}

// todo: _.random to support floats
