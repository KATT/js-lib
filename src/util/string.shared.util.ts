export class StringSharedUtil {
  capitalizeFirstLetter (s: string): string {
    return s.charAt(0).toUpperCase() + s.slice(1)
  }

  lowercaseFirstLetter (s: string): string {
    return s.charAt(0).toLowerCase() + s.slice(1)
  }

  removeWhitespace (s: string): string {
    return s.replace(/\s/g, '')
  }
}

export const stringSharedUtil = new StringSharedUtil()
