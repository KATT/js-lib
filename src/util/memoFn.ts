import { getArgsSignature } from '../decorators/decorator.util'
import { MemoOpts } from '../decorators/memo.decorator'
import { jsonMemoSerializer, MapMemoCache, MemoCache } from '../decorators/memo.util'
import { since } from './time.util'

interface MemoizedFunction {
  cache: MemoCache
}

export function memoFn<T extends (...args: any[]) => any> (
  fn: T,
  opts: MemoOpts = {},
): T & MemoizedFunction {
  const { logHit, logMiss, noLogArgs, cacheFactory, noCacheRejected, noCacheResolved } = {
    cacheFactory: () => new MapMemoCache(),
    ...opts,
  }

  const cache = cacheFactory()
  const awaitPromise = Boolean(noCacheRejected || noCacheResolved)
  const fnName = fn.name

  const memoizedFn = function (this: any, ...args: any[]): T {
    const ctx = this
    const cacheKey = jsonMemoSerializer(args)

    if (cache.has(cacheKey)) {
      if (logHit) {
        console.log(`${fnName}(${getArgsSignature(args, noLogArgs)}) memoFn hit`)
      }

      const res = cache.get(cacheKey)

      if (awaitPromise) {
        return res instanceof Error ? (Promise.reject(res) as any) : Promise.resolve(res)
      } else {
        return res
      }
    }

    const started = Date.now()

    const res: any = fn.apply(ctx, args)

    if (awaitPromise) {
      return (res as Promise<any>)
        .then(res => {
          // console.log('RESOLVED', res)
          if (logMiss) {
            console.log(
              `${fnName}(${getArgsSignature(args, noLogArgs)}) memoFn miss resolved (${since(
                started,
              )})`,
            )
          }

          if (!noCacheResolved) {
            cache.set(cacheKey, res)
          }

          return res
        })
        .catch(err => {
          // console.log('REJECTED', err)
          if (logMiss) {
            console.log(
              `${fnName}(${getArgsSignature(args, noLogArgs)}) memoFn miss rejected (${since(
                started,
              )})`,
            )
          }

          if (!noCacheRejected) {
            // We put it to cache as raw Error, not Promise.reject(err)
            // So, we'll need to check if it's instanceof Error to reject it or resolve
            // Wrap as Error if it's not Error
            cache.set(cacheKey, err instanceof Error ? err : new Error(err))
          }

          return Promise.reject(err)
        }) as any
    } else {
      if (logMiss) {
        console.log(
          `${fnName}(${getArgsSignature(args, noLogArgs)}) @memo miss (${since(started)})`,
        )
      }

      cache.set(cacheKey, res)
      return res
    }
  }

  Object.assign(memoizedFn, { cache })
  return memoizedFn as T & MemoizedFunction
}
