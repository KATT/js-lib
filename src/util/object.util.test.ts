import { deepFreeze } from '../testing/test.shared.util'
import { objectUtil } from './object.util'

test('pick', () => {
  const f = objectUtil.pick.bind(objectUtil)

  expect(f(undefined)).toBe(undefined)
  expect(f(null)).toBe(null)
  expect(f({})).toEqual({})

  const obj = {
    a: 1,
    b: 2,
    c: 3,
    d: false,
  }

  expect(f(obj)).toEqual(obj) // no fields

  const fields: string[] = ['a', 'c', 'd', 'e']
  const r = f(obj, fields)
  // console.log(r)
  expect(r).toEqual({ a: 1, c: 3, d: false })
  expect('e' in r).toBe(false) // should not add more fields with 'undefined' value
  // should not mutate
  expect(obj.c).toBe(3)
})

test('deepTrim', () => {
  const f = objectUtil.deepTrim.bind(objectUtil)
  const o = {
    a: 'abc ',
    b: 'c',
    d: 12,
    e: {
      f: '  sd a ',
    },
  }
  f(o)

  expect(o).toEqual({
    a: 'abc',
    b: 'c',
    d: 12,
    e: {
      f: 'sd a',
    },
  })

  expect(f(undefined)).toBe(undefined)

  expect(
    objectUtil.deepTrim({
      a: ' b',
      c: 'd ',
      e: ' f ',
      g: 'hh',
    }),
  ).toEqual({
    a: 'b',
    c: 'd',
    e: 'f',
    g: 'hh',
  })
})

test('filterFalsyValues', () => {
  const f = objectUtil.filterFalsyValues.bind(objectUtil)
  const o = Object.freeze({
    a: 1,
    b: 0,
    c: undefined,
    d: null,
    e: '',
    f: 'wer',
  })
  expect(f(o)).toEqual({
    a: 1,
    f: 'wer',
  })
  // should NOT mutate
  expect(o.b).toBe(0)

  // should mutate if needed
  const o2 = { ...o }
  f(o2, true)
  expect(o2.b).toBe(undefined)
})

test('filterEmptyStringValues', () => {
  const f = objectUtil.filterEmptyStringValues.bind(objectUtil)
  const o = {
    a: 1,
    b: 0,
    c: undefined,
    d: null,
    e: '',
    f: 'wer',
  }
  expect(f(o)).toEqual({
    a: 1,
    b: 0,
    c: undefined,
    d: null,
    f: 'wer',
  })
  // should NOT mutate
  expect(o.e).toBe('')

  // should mutate if needed
  f(o, true)
  expect(o.e).toBe(undefined)
})

test('filterUndefinedValues', () => {
  const o = {
    a: 1,
    b: 0,
    c: undefined,
    d: null,
    e: '',
    f: 'wer',
  }
  deepFreeze(o)

  expect(objectUtil.filterUndefinedValues(o)).toEqual({
    a: 1,
    b: 0,
    e: '',
    f: 'wer',
  })
})

test('filterValues', () => {
  const f = objectUtil.filterValues.bind(objectUtil)
  const br = {
    a: 'b',
    c: null,
  }

  // should mutate
  f(br, (k: any, v: any) => v !== null, true)
  expect(br.c).toBeUndefined()
})

test('objectNullValuesToUndefined', () => {
  const o = {
    a: undefined,
    b: null,
    c: 1,
  }
  expect(objectUtil.objectNullValuesToUndefined(o)).toEqual({
    a: undefined,
    b: undefined,
    c: 1,
  })
})

test('deepEquals Issue!', () => {
  expect(
    objectUtil.deepEquals(
      {
        a: 1,
        b: 2,
      },
      {
        b: 2,
        a: 1,
      },
    ),
  ).toBe(true)
})

test('deepEquals', () => {
  expect(
    objectUtil.deepEquals(
      {
        a: 1,
        b: 2,
      },
      {
        a: 1,
        b: 2,
      },
    ),
  ).toBe(true)

  expect(
    objectUtil.deepEquals(
      {
        a: 1,
        b: 2,
      },
      {
        a: 1,
        b: 2,
        c: 3,
      },
    ),
  ).toBe(false)
})

test('unsetValue', () => {
  const o = {
    a: 1,
    b: {
      c: 3,
      d: 4,
    },
    m: 5,
  }
  objectUtil.unsetValue(o, 'b.c')
  objectUtil.unsetValue(o, 'm')
  expect(o).toEqual({
    a: 1,
    b: {
      d: 4,
    },
  })
  expect(o.m).toBeUndefined()
  expect(o.b.c).toBeUndefined()
})

test('deepCopy', () => {
  const o = {
    a: 1,
    b: {
      c: 3,
      d: 4,
    },
  }
  expect(objectUtil.deepCopy(o)).toEqual(o)
})

test('isObject', () => {
  expect(objectUtil.isObject(undefined)).toBe(false)
})

test('isEmptyObject', () => {
  const a = [1, 0, -1, undefined, null, 'wer', 'a', '', {}, { a: 'b' }]

  const empty = a.filter(i => objectUtil.isEmptyObject(i))
  expect(empty).toEqual([{}])
})

test('mergeDeep', () => {
  const a1 = {
    b: {
      c: 'c1',
    },
    e: 'e1',
  }
  const a2 = {
    b: {
      c2: 'c2',
    },
    d: 'd2',
  }

  expect(objectUtil.mergeDeep(a1, a2)).toMatchSnapshot()
})

test('mask', () => {
  const o = {
    a: '1',
    b: {
      c: '2',
      d: '1',
      e: {
        f: 7,
      },
    },
  }
  const r = objectUtil.mask(o, ['b.c'])
  expect(r).toEqual({
    a: '1',
    b: {
      d: '1',
      e: {
        f: 7,
      },
    },
  })
  expect(r.b.c).toBeUndefined()
  expect(r.b.e !== o.b.e)

  // should not fail
  expect(objectUtil.mask(o, ['c.0.0'])).toEqual(o)
})

test('sortObjectDeep', () => {
  const o = {
    b: 2,
    a: 1,
    c: {
      d: {
        m: 4,
        j: 5,
        aa: 1,
      },
      m: 'sdf',
      dd: [3, 2, 1, 0, 55],
      ddd: ['b', 'c', 'a'],
    },
    aa: 6,
  }

  // console.log(objectUtil.sortObjectDeep(o))
  expect(objectUtil.sortObjectDeep(o)).toMatchSnapshot()
})

test('arrayToHash', () => {
  const f = objectUtil.arrayToHash

  expect(f(undefined)).toEqual({})
  expect(f([])).toEqual({})

  const a = ['a', 'b']
  expect(f(a)).toEqual({
    a: true,
    b: true,
  })
})

test('classToPlain', () => {
  const f = objectUtil.classToPlain

  class A {
    constructor (public a?: string, public b?: string) {}
  }

  expect(f(undefined)).toEqual({})
  expect(f(null)).toEqual(null)
  expect(f({})).toEqual({})
  expect(f(new A('aa', 'bb'))).toEqual({ a: 'aa', b: 'bb' })
})

test('getKeyByValue', async () => {
  const o = {
    a: 'ak',
    b: 'bk',
  }
  expect(objectUtil.getKeyByValue(o, 'bk')).toBe('b')
  expect(objectUtil.getKeyByValue(o, 'ak')).toBe('a')
})

test('invertObject', async () => {
  const o = {
    a: 'ak',
    b: 'bk',
  }
  const inv = {
    ak: 'a',
    bk: 'b',
  }

  expect(objectUtil.invertObject(o)).toEqual(inv)
})

test('invertMap', async () => {
  const o = new Map([['a', 'ak'], ['b', 'bk']])
  const inv = new Map([['ak', 'a'], ['bk', 'b']])

  expect(objectUtil.invertMap(o)).toEqual(inv)
})

test('by', () => {
  expect(objectUtil.by(undefined, 'a')).toEqual({})

  const a = [{ a: 'aa' }, { a: 'ab' }, { b: 'bb' }]
  const r = objectUtil.by(a, 'a')
  expect(r).toEqual({
    aa: { a: 'aa' },
    ab: { a: 'ab' },
  })
})
