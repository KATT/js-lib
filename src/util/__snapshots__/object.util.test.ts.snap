// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeDeep 1`] = `
Object {
  "b": Object {
    "c": "c1",
    "c2": "c2",
  },
  "d": "d2",
  "e": "e1",
}
`;

exports[`sortObjectDeep 1`] = `
Object {
  "a": 1,
  "aa": 6,
  "b": 2,
  "c": Object {
    "d": Object {
      "aa": 1,
      "j": 5,
      "m": 4,
    },
    "dd": Array [
      3,
      2,
      1,
      0,
      55,
    ],
    "ddd": Array [
      "b",
      "c",
      "a",
    ],
    "m": "sdf",
  },
}
`;
